File: myrm-agent-server/app/api/chats/chat/copilot.py
"""Co-Pilot API — run digest + session advisor.

[INPUT]
HTTP requests for run digest CRUD and advisor chat.

[OUTPUT]
JSON responses with digest data or advisor thread messages.

[POS]
Thin API layer delegating to copilot service modules.
"""

from __future__ import annotations

from fastapi import APIRouter, Request
from pydantic import BaseModel, Field

from app.core.utils.response_utils import error_response, success_response
from app.services.copilot.advisor_service import ask_advisor
from app.services.copilot.advisor_thread_store import AdvisorThreadStore
from app.services.copilot.run_digest_store import RunDigestStore

router = APIRouter()


class AdvisorAskRequest(BaseModel):
    question: str = Field(..., min_length=1, max_length=4000)
    selection_snippet: str | None = Field(default=None, max_length=2000)


@router.get("/{chat_id}/copilot/run-digest")
async def get_run_digest(chat_id: str):
    digest = RunDigestStore.get(chat_id)
    if digest is None:
        return success_response(data={"digest": None})
    return success_response(data={"digest": digest.to_dict()})


@router.get("/{chat_id}/copilot/advisor/messages")
async def list_advisor_messages(chat_id: str):
    messages = [msg.to_dict() for msg in AdvisorThreadStore.list_messages(chat_id)]
    return success_response(data={"messages": messages})


@router.delete("/{chat_id}/copilot/advisor/messages")
async def clear_advisor_messages(chat_id: str):
    AdvisorThreadStore.clear(chat_id)
    return success_response(data={"cleared": True})


@router.post("/{chat_id}/copilot/advisor/ask")
async def advisor_ask(chat_id: str, body: AdvisorAskRequest, request: Request):
    question = body.question.strip()
    if not question:
        return error_response(message="Question cannot be empty", code=400)

    accept_lang = request.headers.get("Accept-Language", "en")

    AdvisorThreadStore.append(parent_chat_id=chat_id, role="user", content=question)
    reply, tier = await ask_advisor(
        chat_id=chat_id,
        question=question,
        selection_snippet=body.selection_snippet,
        accept_language=accept_lang,
    )
    msg = AdvisorThreadStore.append(
        parent_chat_id=chat_id,
        role="assistant",
        content=reply,
        tier=tier,
    )
    return success_response(
        data={
            "reply": reply,
            "tier": tier,
            "message": msg.to_dict(),
        }
    )

File: myrm-agent-frontend/src/services/copilot.ts
import { fetchWithTimeout } from '@/lib/api';

export interface RunDigestStep {
  index: number;
  tool_name: string;
  step_key: string;
  status?: string | null;
}

export interface RunDigest {
  chat_id: string;
  phase: 'idle' | 'running' | 'waiting_approval' | 'completed' | 'error' | 'cancelled';
  step_count: number;
  current_tool: string | null;
  current_step_key: string | null;
  pending_approval_count: number;
  elapsed_seconds: number;
  headline: string;
  recent_steps: RunDigestStep[];
  updated_at: string;
}

export interface AdvisorMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  tier: string;
  created_at: string;
}

export async function fetchRunDigest(chatId: string): Promise<RunDigest | null> {
  const res = await fetchWithTimeout(`/agents/chats/${chatId}/copilot/run-digest`);
  if (!res.ok) return null;
  const json = (await res.json()) as { data?: { digest?: RunDigest | null } };
  return json.data?.digest ?? null;
}

export async function askAdvisor(
  chatId: string,
  question: string,
  selectionSnippet?: string,
): Promise<{ reply: string; tier: string; message: AdvisorMessage } | null> {
  const res = await fetchWithTimeout(`/agents/chats/${chatId}/copilot/advisor/ask`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      question,
      selection_snippet: selectionSnippet ?? null,
    }),
  });
  if (!res.ok) return null;
  const json = (await res.json()) as {
    data?: { reply: string; tier: string; message: AdvisorMessage };
  };
  if (!json.data) return null;
  return json.data;
}

export async function fetchAdvisorMessages(chatId: string): Promise<AdvisorMessage[]> {
  const res = await fetchWithTimeout(`/agents/chats/${chatId}/copilot/advisor/messages`);
  if (!res.ok) return [];
  const json = (await res.json()) as { data?: { messages?: AdvisorMessage[] } };
  return json.data?.messages ?? [];
}

export async function clearAdvisorMessages(chatId: string): Promise<void> {
  await fetchWithTimeout(`/agents/chats/${chatId}/copilot/advisor/messages`, {
    method: 'DELETE',
  });
}
